### PR TITLE
Implement redis connection broken mode

### LINKED ISSUE #569: poor default for retry_max_delay
doubling without a cap is not reasonable IMO, even a large cap would be a much saner default

### PR DESCRIPTION
**Pull request by BridgeAR**
This is also going to change some small things like adding a maximum timeout.

### CODE REVIEW

**Comment by BridgeAR** [comment]
This branch is going to fix a lot of issues for quite some people.

**Comment by BridgeAR** [comment]
Fixes #569, #587, #566 and #280.
This includes the fix as suggested in #671 and #533
Closes #675, #615, #463 and #724

There will be more closed tickets that we closed without fixing that get fixed by this PR.

File: index.js [hunk 1]
```diff
@@ -1063,7 +1051,7 @@ Multi.prototype.exec = function (callback) {
 
     // TODO - make this callback part of Multi.prototype instead of creating it each time
     return this._client.send_command("exec", [], function (err, replies) {
-        if (err) {
+        if (err && !err.code) {
```
Comment (BridgeAR) [inline index.js]: This is a preparation for the multi fixes.

File: .jshintrc [hunk 2]
```diff
@@ -16,6 +16,9 @@
     "node": true, // Enable globals available when code is running inside of the NodeJS runtime environment.
     "mocha": true,
 
+    // Relaxing options
+    "boss": true, // Accept statements like `while (key = keys.pop()) {}`
```
Comment (BridgeAR) [inline .jshintrc]: I personally like to use this. I could imagine though that not everyone likes it?

File: README.md [hunk 3]
```diff
@@ -188,10 +193,11 @@ with an error, or an error will be thrown if no callback is specified.
...
 reconnection (delay) almost doubles. This delay normally grows infinitely, but setting `retry_max_delay` limits delay
 to maximum value, provided in milliseconds.
-* `connect_timeout` defaults to `false`. By default client will try reconnecting until connected. Setting `connect_timeout`
-limits total time for client to reconnect. Value is provided in milliseconds and is counted once the disconnect occured.
-* `max_attempts` defaults to `null`. By default client will try reconnecting until connected. Setting `max_attempts`
-limits total amount of reconnects.
+* `connect_timeout` defaults to `86400000`. Setting `connect_timeout` limits total time for client to reconnect.
+Value is provided in milliseconds and is counted once the disconnect occured. The last retry is going to happen exactly at the timeout time.
+That way the default is to try reconnecting until 24h passed.
+* `max_attempts` defaults to `0`. By default client will try reconnecting until connected. Setting `max_attempts`
```
Comment (BridgeAR) [inline README.md]: The new default should not break any old code, since `null` would behave just the same as `0`.

File: README.md [hunk 4]
```diff
@@ -188,10 +193,11 @@ with an error, or an error will be thrown if no callback is specified.
 * `retry_max_delay`: defaults to `null`. By default every time the client tries to connect and fails time before
 reconnection (delay) almost doubles. This delay normally grows infinitely, but setting `retry_max_delay` limits delay
 to maximum value, provided in milliseconds.
-* `connect_timeout` defaults to `false`. By default client will try reconnecting until connected. Setting `connect_timeout`
-limits total time for client to reconnect. Value is provided in milliseconds and is counted once the disconnect occured.
-* `max_attempts` defaults to `null`. By default client will try reconnecting until connected. Setting `max_attempts`
-limits total amount of reconnects.
+* `connect_timeout` defaults to `86400000`. Setting `connect_timeout` limits total time for client to reconnect.
```
Comment (BridgeAR) [inline README.md]: This could be split into another PR but I'd like to keep it all together

File: README.md [hunk 5]
```diff
@@ -575,12 +581,12 @@ some kind of maximum queue depth for pre-connection commands.
 
 ## client.retry_delay
 
-Current delay in milliseconds before a connection retry will be attempted. This starts at `250`.
+Current delay in milliseconds before a connection retry will be attempted. This starts at `200`.
```
Comment (BridgeAR) [inline README.md]: I think starting with a lower delay might be good and we might actually also consider lowering the multiplier. What do you all think?

File: test/auth.spec.js [hunk 6]
```diff
@@ -12,10 +12,11 @@ describe("client authentication", function () {
         });
     });
 
-    helper.allTests(function(parser, ip, args) {
+    helper.allTests({
+        allConnections: true
+    }, function(parser, ip, args) {
```
Comment (BridgeAR) [inline test/auth.spec.js]: The test suite has a new default to only run each test with each parser once instead of each parser * each connection type (IPv4, IPv6 and socket). This reduces unnecessary overhead. To run each connection type this optional option may be passed to the helper.

File: README.md (hunk 4 above)
Comment (BridgeAR) [inline README.md]: A whole day as a maximum delay might already be to much.

**Comment by BridgeAR** [comment]
This is going to do the following:

1. [FIX?] Change the default connection timeout to be one day instead of never.
2. [FIX] There won't be any connection retries after the connection timeout anymore. This was the case earlier because the timeout for the next retry is increasing each retry by a multiplier. If the next retry would be above the connection timeout it is now going to be exactly at the maximum retry connection timeout instead. So the timeout is follows as expected from now on. There
3. [??] The default minimum connection retry time is reduced a tiny bit (200 ms instead of 255).
4. [FIX] The queues won't be flushed anymore if node_redis is trying to reconnect. It will only be flushed if the connection is broken or if the user closes the connection. So offline_queue and max_attempts / connection_timeout is now properly working together.
5. [FEATURE] If no connection could be established and max_attempts / connection_timeout has been exceeded node_redis is going into the "connection is broken" mode and emit an error.

File: README.md (hunk 3 above)
Comment (bcoe) [inline README.md]: `0` seems a bit weird to me, to me this feels like:

`don't try to reconnect`.

How would I tell the client not to reconnect?

Comment (BridgeAR) [inline README.md]: I thought so too. But this is actually just a stylistic change, since passing 0 earlier behaved just the same, it was just not documented. So I thought it's probably best to change `null` to `0` and document it.

> How would I tell the client not to reconnect?

Have a look at the next line :D (entering 1 does that - it's about the connection attempts in total not the reconnect attempts. While fixing the max_attempts I first switched it to be the number of reconnects and therefor `0` would have been `don't try to reconnect` but this is not represented by the variable name and did not behave like that before. So this seems to be legit).

**Comment by BridgeAR** [comment]
This should be ready to merge
